src/workerd/io/actor-sqlite-test.c++

```diff
@@ -631,6 +631,62 @@ KJ_TEST("canceling deferred alarm deletion is idempotent") {
   KJ_ASSERT(expectSync(test.getAlarm()) == oneMs);
 }
 
+KJ_TEST("alarm handler cleanup succeeds when output gate is broken and handler fails") {
+  ActorSqliteTest test({.monitorOutputGate = false});
+  auto promise = test.gate.onBroken();
+
+  // Initialize alarm state to 1ms.
+  test.setAlarm(oneMs);
+  test.pollAndExpectCalls({"scheduleRun(1ms)"})[0]->fulfill();
+  test.pollAndExpectCalls({"commit"})[0]->fulfill();
+  test.pollAndExpectCalls({});
+  KJ_ASSERT(expectSync(test.getAlarm()) == oneMs);
+
+  auto armResult = test.actor.armAlarmHandler(oneMs, false);
+  KJ_ASSERT(armResult.is<ActorSqlite::RunAlarmHandler>());
+
+  // Break gate
+  test.put("foo", "bar");
+  test.pollAndExpectCalls({"commit"})[0]->reject(KJ_EXCEPTION(FAILED, "a_rejected_commit"));
+  KJ_EXPECT_THROW_MESSAGE("a_rejected_commit", promise.wait(test.ws));
+  // Ensure taskFailed handler runs and notices brokenness:
+  test.ws.poll();
+
+  // In the case where the handler fails, we assume the caller will explicitly cancel deferred
+  // alarm deletion:
+  test.actor.cancelDeferredAlarmDeletion();
+
+  // Dropping the DeferredAlarmDeleter should succeed.
+  { auto drop = kj::mv(armResult); }
+}
+
+KJ_TEST("alarm handler cleanup succeeds when output gate is broken but handler succeeds") {
+  ActorSqliteTest test({.monitorOutputGate = false});
+  auto promise = test.gate.onBroken();
+
+  // Initialize alarm state to 1ms.
+  test.setAlarm(oneMs);
+  test.pollAndExpectCalls({"scheduleRun(1ms)"})[0]->fulfill();
+  test.pollAndExpectCalls({"commit"})[0]->fulfill();
+  test.pollAndExpectCalls({});
+  KJ_ASSERT(expectSync(test.getAlarm()) == oneMs);
+
+  auto armResult = test.actor.armAlarmHandler(oneMs, false);
+  KJ_ASSERT(armResult.is<ActorSqlite::RunAlarmHandler>());
+
+  // Break gate
+  test.put("foo", "bar");
+  test.pollAndExpectCalls({"commit"})[0]->reject(KJ_EXCEPTION(FAILED, "a_rejected_commit"));
+  KJ_EXPECT_THROW_MESSAGE("a_rejected_commit", promise.wait(test.ws));
+  // Ensure taskFailed handler runs and notices brokenness:
+  test.ws.poll();
```

Comment:
Up to and including this line, this test is the same as the test above it.  Why does it mean the handler succeeds in this case, but fails in the previous one?

Reply:
During normal alarm processing, if the alarm handler fails (i.e. the application throws an exception), `cancelDeferredAlarmDeletion()` is invoked.

Reply:
Right...  The intent behind the two tests was to exercise both the case where the caller calls `cancelDeferredAlarmDeletion()` prior to dropping the deleter and the case where it doesn't.  Strictly speaking, only the behavior of the non-cancelled test changes in this commit; the cancelled test was already succeeding because it didn't attempt to reset the alarm after brokenness, but I don't think the interaction between gate brokenness and cancellation was otherwise being covered by other tests.

- [ ] I can try to clarify the comments.